### PR TITLE
Refactor Tabulator to read and write in separate pools

### PR DESCRIPTION
**Pull request by chases2**
Follow-up to https://github.com/GoogleCloudPlatform/testgrid/pull/942

### CODE REVIEW

**Comment by chases2** [comment]
/assign @fejta 
Please take a look when you are able!

File: pkg/tabulator/tabstate.go [hunk 1]
```diff
@@ -183,80 +197,133 @@ func Update(ctx context.Context, client gcs.ConditionalClient, mets *Metrics, co
...
+			q.Fix(group.Name, time.Now().Add(freq/10), false)
+			return nil
+		}
+
+		fromPath, err := updater.TestGroupPath(configPath, gridPathPrefix, group.Name)
+		if err != nil {
+			return fmt.Errorf("can't make tg path %q: %w", group.Name, err)
+		}
+
+		log.WithFields(logrus.Fields{
```
Comment (fejta) [inline pkg/tabulator/tabstate.go:221]: `log.WithField("from", fromPath)` as you're only using a single field.

File: pkg/tabulator/tabstate.go [hunk 2]
```diff
@@ -183,80 +197,133 @@ func Update(ctx context.Context, client gcs.ConditionalClient, mets *Metrics, co
...
-			}).Info("Calculating state")
-
-			err = createTabState(ctx, log, client, view.tab, group, *fromPath, *toPath, confirm, dropEmptyCols, calculateStats, useTabAlertSettings)
-			if err != nil {
-				return fmt.Errorf("can't calculate state: %w", err)
-			}
+			out := view
+			out.data = proto.Clone(grid).(*statepb.Grid)
+			out.done = wg.Done
+			tabs <- out
```
Comment (fejta) [inline pkg/tabulator/tabstate.go:238]: Shouldn't this pay attention to when the context expires?

File: pkg/tabulator/tabstate.go [hunk 3]
```diff
@@ -183,80 +197,133 @@ func Update(ctx context.Context, client gcs.ConditionalClient, mets *Metrics, co
...
-				"to":   toPath.String(),
-			}).Info("Calculating state")
-
-			err = createTabState(ctx, log, client, view.tab, group, *fromPath, *toPath, confirm, dropEmptyCols, calculateStats, useTabAlertSettings)
-			if err != nil {
-				return fmt.Errorf("can't calculate state: %w", err)
-			}
+			out := view
+			out.data = proto.Clone(grid).(*statepb.Grid)
+			out.done = wg.Done
```
Comment (fejta) [inline pkg/tabulator/tabstate.go:237]: hmm, are you sure about this? Seems fragile

File: pkg/tabulator/tabstate.go [hunk 4]
```diff
@@ -183,80 +197,133 @@ func Update(ctx context.Context, client gcs.ConditionalClient, mets *Metrics, co
...
-
-				log := log.WithField("group", group.Name)
+			}
+		}()
+	}
+	writeWg.Add(writeConcurrency)
+	for i := 0; i < writeConcurrency; i++ {
+		go func() {
+			defer writeWg.Done()
+			for tab := range tabs {
```
Comment (fejta) [inline pkg/tabulator/tabstate.go:271]: Would task be a better name here?

File: pkg/tabulator/tabstate.go [hunk 5]
```diff
@@ -183,80 +197,133 @@ func Update(ctx context.Context, client gcs.ConditionalClient, mets *Metrics, co
...
-				"from": fromPath.String(),
-				"to":   toPath.String(),
-			}).Info("Calculating state")
-
-			err = createTabState(ctx, log, client, view.tab, group, *fromPath, *toPath, confirm, dropEmptyCols, calculateStats, useTabAlertSettings)
-			if err != nil {
-				return fmt.Errorf("can't calculate state: %w", err)
-			}
+			out := view
+			out.data = proto.Clone(grid).(*statepb.Grid)
```
Comment (fejta) [inline pkg/tabulator/tabstate.go:236]: Seems a waste of inflate N times? Isn't this what we spend most of our CPU time doing?

**Review by fejta — APPROVED** [review]
/hold

File: pkg/tabulator/tabstate.go (hunk 5 above)
Comment (chases2) [reply pkg/tabulator/tabstate.go]: Not sure what you mean. Deep-copying the result is cheaper than inflating it each time.

File: pkg/tabulator/tabstate.go (hunk 4 above)
Comment (chases2) [reply pkg/tabulator/tabstate.go]: Good point; renamed

File: pkg/tabulator/tabstate.go (hunk 2 above)
Comment (chases2) [reply pkg/tabulator/tabstate.go]: Added a skip here

File: pkg/tabulator/tabstate.go (hunk 5 above)
Comment (fejta) [reply pkg/tabulator/tabstate.go]: This code downloads the proto once and then inflates it N times.

File: pkg/tabulator/tabstate.go (hunk 3 above)
Comment (fejta) [reply pkg/tabulator/tabstate.go]: WaitGroup seems fine, but passing a wait group to a foreign function via a channel seems fragile and hard to reason about.

Comment (fejta) [reply pkg/tabulator/tabstate.go]: (I want be able to read a single function to validate that `wg.Done()` is called `wg.Add(n)` times)

File: pkg/tabulator/tabstate.go (hunk 5 above)
Comment (chases2) [reply pkg/tabulator/tabstate.go]: Ah, yes, of course it does. I have no idea why I didn't see that before. I'll look into passing inflated columns or something similar

File: pkg/tabulator/tabstate.go (hunk 3 above)
Comment (fejta) [reply pkg/tabulator/tabstate.go]: Maybe instead of passing a wg.Done why not have the writer remove the item from the processing set?

Comment (chases2) [reply pkg/tabulator/tabstate.go]: The writer might not be the only one writing a test group; test groups to tabs are one-to-many. The processing set could be dashboard#tabs instead, though. 🤔

Comment (fejta) [reply pkg/tabulator/tabstate.go]: Happy to go with this and wait until there's empirical evidence from debugging a mysterious "why does the tabulator stop doing anything after awhile?" issue after a refactor to conclude that this is a fragile pattern.

File: pkg/tabulator/tabstate.go (hunk 5 above)
Comment (chases2) [reply pkg/tabulator/tabstate.go]: I can't reasonably fix this until "--filter-columns" is dropped as a feature flag. Will have to address this in a later PR.

File: pkg/tabulator/tabstate.go (hunk 3 above)
Comment (chases2) [reply pkg/tabulator/tabstate.go]: I thought a bit more about _why_ we want to call back to the reading thread like this at all, and I'm not convinced we need it. These routines are already architected to pass off everything before another reader gets to pass off anything.

Deleting code is a pretty good way to make it simpler, so I would favor that solution

**Review by fejta — APPROVED** [review]

**Pull request merged**